### ISSUE TITLE
printshellcmnds does not work

### ISSUE THREAD

**Issue by trevorcousins** [snakemake/snakemake#2090]
The --printshellcmnds is supposed to print out the shell commands from each rule. However it does not do anything and I believe this is a bug. 
Tried for version 7.20.0 and 7.15.2, both failed. 
According to stackoverflow this seems to have been a problem since 2019.
https://stackoverflow.com/questions/47415371/is-it-possible-to-print-commands-instead-of-rules-in-snakemake-dry-run

Is there anyway around this?

**Comment by dariober** [comment]
Can you show an example? I use `-p` all the time and it works.

**Comment by trevorcousins** [comment]
Well, it doesn't work for any of my snakefiles but here is a minimal example.

Snakefile is

```
files = range(1,3)
rule all:
        input:
                [f'/home/tc557/snakemake_example_230209/generated_data/file{z}.txt' for z in files]

rule generate_data:
        output:
                outfile = '/home/tc557/snakemake_example_230209/generated_data/file{z}.txt'
        run:
                shell(
                        f'python generate_data.py -parameters a b c -o {output.outfile}'
                )
```

When I run `snakemake -n` I get 


```
Building DAG of jobs...
Job stats:
job              count    min threads    max threads
-------------  -------  -------------  -------------
all                  1              1              1
generate_data        2              1              1
total                3              1              1


[Thu Feb  9 12:23:55 2023]
rule generate_data:
    output: /home/user/snakemake_example_230209/generated_data/file1.txt
    jobid: 1
    reason: Missing output files: /home/user/snakemake_example_230209/generated_data/file1.txt
    wildcards: z=1
    resources: tmpdir=/tmp

[Thu Feb  9 12:23:55 2023]
rule generate_data:
    output: /home/user/snakemake_example_230209/generated_data/file2.txt
    jobid: 2
    reason: Missing output files: /home/user/snakemake_example_230209/generated_data/file2.txt
    wildcards: z=2
    resources: tmpdir=/tmp

[Thu Feb  9 12:23:55 2023]
localrule all:
    input: /home/user/snakemake_example_230209/generated_data/file1.txt, /home/user/snakemake_example_230209/generated_data/file2.txt
    jobid: 0
    reason: Input files updated by another job: /home/user/snakemake_example_230209/generated_data/file1.txt, /home/user/snakemake_example_230209/generated_data/file2.txt
    resources: tmpdir=/tmp

Job stats:
job              count    min threads    max threads
-------------  -------  -------------  -------------
all                  1              1              1
generate_data        2              1              1
total                3              1              1

Reasons:
    (check individual jobs above for details)
    input files updated by another job:
        all
    missing output files:
        generate_data

This was a dry-run (flag -n). The order of jobs does not reflect the order of execution.
```

And when I run `snakemake -n -p` I get exactly the same. 

I tried with snakemake version 7.15.2 and  7.20.0. Same for each.

**Comment by dariober** [comment]
`--printshellcmds/-p` prints the commands from the `shell` *directive* with or without `-n` execution. However, the commands in a `shell` *function* inside a `run` directive can be printed only in real execution mode, i.e. without `-n`. 

I think this behavior is inevitable because the string you pass to the `shell` function can be prepared only if you run the code for real. Consider this example:

```
rule one:
    input: ...
    output: ...
    run:
        import random

        r = random.random()

        with open('foo', 'w') as fout:
            fout.write(str(r))

        cmd = 'some command %s' % r 
        shell(cmd)
```

If you execute this code with `-n` and you wanted `shell` to print the command it executes, you need to actually execute the python code that comes before it, including writing file `foo`. This would defeat the point of `-n`.

TLDR: The example you posted is the expected behavior and I don't think it can be changed.

**Comment by trevorcousins** [comment]
Oh, I see now. Thanks. I did not realise that you could do 
`shell:
    "some_command`
I thought one had to do 
```
run:
    shell(f'some_command')
```

I'm not sure why I thought it was standard to use the `run: shell()` syntax. Are there ever advantages of doing that?

**Comment by dariober** [comment]
I'm not aware of anything that you can do with the `shell:` directive that you cannot do with the combination `run: shell(...)`. I would say the behavior of `-p -n` is the main difference but that's an important one since it is incredibly useful.

PS: I think you can close this issue?

**Comment by trevorcousins** [comment]
Ok thanks. One more question. Often the line that I want to execute for the shell involves functions of wildcards and variables from elsewhere in the Snakefile script. For example here is Snakefile similar to what I use:
```

rule run_algorithm:
    input:
        files = get_input_files,
    output:
        outfile = '/path/to/outfile/final_run.txt'    
    run:        
        shell(
            f'python /path/to/algorithm -in {" ".join(input.files)} -o {output.outfile.split("final")[0]} -param1 {wildcards.param1} -param2 {predefinedvariable[wildcards.param2]}'                
        )
```
`get_input_files`  is a function whose only argument is wildcards. I split the `outfile` because the `-o` flag is the output dir and the algorithm outputs the file "final_run.txt" on top of the directory. I need to use `predefinedvariable` as opposed to writing that in as a wildcard for intepretability. 
Writing this in shell directive does not seem to be possible. Any ideas?

**Comment by LukaP-BB** [comment]
With a bit of refactoring you could do something like, assuming `predefinedvariable` is a dictionary  : 

```
rule run_algorithm:
    input:
        files = get_input_files,
    output:
        outfile = '/path/to/outfile/final_run.txt'    
    params:
       out_folder='/path/to/outfile/',
       param2 = lambda wildcards : predefinedvariable[wildcards.param2]
    shell:
        """
        python /path/to/algorithm -in {input.files} -o {params.out_folder} -param1 {wildcards.param1} -param2 {params.param2}
       """
```
Your method above is really convoluted : you're using snakemake which is basically a python program, to call a subshell which is itself calling another python subprocess. It works but that's not the most beautiful.

Also, you coul import the main function of your algorithm and run it directly within the run directive, eg : 

```
from my_algorithm import main_function

rule run_algorithm:
    input:
        files = get_input_files,
    output:
        outfile = '/path/to/outfile/final_run.txt'    
    run:
        main_function(all, necessary, parameters)
```

**Comment by trevorcousins** [comment]
Yes I agree my method is a bit ugly, but that was the only way I could get it to work. Thanks for your suggestion, that looks good. I will close this now. Thanks for the help.
